app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java

```diff
@@ -680,18 +663,75 @@ public boolean onBackPressed() {
     // Only supports java.io.File. TODO: Android Content
     public boolean saveDocument() {
         boolean ret = false;
-        if (isAdded() && _hlEditor != null && _hlEditor.getText() != null) {
+        if (isAdded() && _document != null && _hlEditor != null && _hlEditor.getText() != null) {
             ret = DocumentIO.saveDocument(_document, _hlEditor.getText().toString(), _shareUtil, getContext());
             updateLauncherWidgets();
 
-            if (_document != null && _document.getFile() != null) {
-                _appSettings.setLastEditPosition(_document.getFile(), _hlEditor.getSelectionStart(), _hlEditor.getTop());
-                _appSettings.setDocumentPreviewState(getPath(), _isPreviewVisible);
-            }
+            _appSettings.setDocumentPreviewState(getPath(), _isPreviewVisible);
+            saveDocumentPositions();
         }
         return ret;
     }
 
+    public void saveDocumentPositions() {
+        if (!Arrays.asList(_hlEditor, _webView, _document.getFile()).contains(null)) {
+            if (_isPreviewVisible) {
+                _appSettings.setLastViewPosition(_document.getFile(), _webView.getScrollX(), _webView.getScrollY());
+            } else {
+                _appSettings.setLastEditPosition(_document.getFile(), _hlEditor.getSelectionStart());
+            }
+        }
+    }
+
+    public void restoreDocumentPositions() {
+        if (!Arrays.asList(_hlEditor, _webView, _document.getFile()).contains(null) && !isTabTodoOrQuickNote()) {
+            int v;
+            if ((v = _document.getInitialLineNumber()) >= 0) { // If Intent contains line number, jump to it
+                _hlEditor.smoothMoveCursorToLine(v);
+            } else { // otherwise take settings
+                if (_isPreviewVisible) {
+                    _webView.scrollAnimatedToXY(_appSettings.getLastViewPositionX(_document.getFile()), _appSettings.getLastViewPositionY(_document.getFile()));
+                    hideSoftKeyboard();
+                } else {
+                    if ((v = _appSettings.isEditorStartOnBottom() || isDocTodoOrQuickNote() ? _hlEditor.length() : _appSettings.getLastEditPositionChar(_document.getFile())) >= 0) {
+                        final int selPosition = v;
+                        _hlEditor.postDelayed(() -> {
+                            if (!_hlEditor.hasFocus()) {
+                                _hlEditor.requestFocus();
+                            }
+                            _hlEditor.setSelection(selPosition);
+                            new ActivityUtils(getActivity()).showSoftKeyboard().freeContextRef();
```

Comment:
So far, i have not found any issues in my testing this latest version. Only some cleanup needed if I can't find more behavior bugs. :+1:

Reply:
@adelobosko I do not think that this postDelayed is actually necessary. `restoreDocumentPositions` is called in `onResume`. UI actions should be fine there without any delay. Works fine on my device without `postDelayed`

Reply:
> 
> 
> @adelobosko I do not think that this postDelayed is actually necessary. `restoreDocumentPositions` is called in `onResume`. UI actions should be fine there without any delay. Works fine on my device without `postDelayed`

@harshad1 hmm, for me delay is needed. I tried without the delay but the selection could not set. Try to open a big file. I think need another way to set selection. After the editor text is loading... i going to remove restore selection in this pull

Reply:
> After the editor text is loading...

But seems 300 milliseconds is enough for loading a big file (selection is set). Need to try a very big file :)